lib/AST/Module.cpp

```diff
@@ -1024,6 +1049,53 @@ void ModuleDecl::getTopLevelDeclsWhereAttributesMatch(
   FORWARD(getTopLevelDeclsWhereAttributesMatch, (Results, matchAttributes));
 }
 
+void ModuleDecl::lookupTopLevelDeclsByObjCName(SmallVectorImpl<Decl *> &Results,
+                                               DeclName name) {
+  if (!isObjCNameLookupCachePopulated())
+    populateObjCNameLookupCache();
+
+  if (name.isSpecial())
+    return;
+
+  auto resultsForFileUnit = ObjCNameLookupCache.find(name.getBaseIdentifier());
+  if (resultsForFileUnit == ObjCNameLookupCache.end())
+    return;
+
+  Results.append(resultsForFileUnit->second.begin(),
+                 resultsForFileUnit->second.end());
+}
+
+void ModuleDecl::populateObjCNameLookupCache() {
+  SmallVector<Decl *> topLevelObjCExposedDeclsInFileUnit;
+  auto hasObjCAttrNamePredicate = [](const DeclAttributes &attrs) -> bool {
+    return attrs.hasAttribute<ObjCAttr>();
+  };
+
+  for (FileUnit *file : getFiles()) {
+    file->getTopLevelDeclsWhereAttributesMatch(
+        topLevelObjCExposedDeclsInFileUnit, hasObjCAttrNamePredicate);
+    if (auto *synth = file->getSynthesizedFile()) {
+      synth->getTopLevelDeclsWhereAttributesMatch(
+          topLevelObjCExposedDeclsInFileUnit, hasObjCAttrNamePredicate);
+    }
+  }
+
+  for (Decl *decl : topLevelObjCExposedDeclsInFileUnit) {
+    if (ValueDecl *VD = dyn_cast<ValueDecl>(decl)) {
+      if (VD->hasName()) {
+        const auto &declObjCAttribute = VD->getAttrs().getAttribute<ObjCAttr>();
+        const Identifier &declObjCName =
+            declObjCAttribute->hasName()
+                ? declObjCAttribute->getName()->getSimpleName()
```

Comment:
Are there no top level decls with non simple names? Existing implementation seems to imply not, but I wasn't sure: https://github.com/apple/swift/blob/main/lib/ClangImporter/ImportDecl.cpp#L4282

Reply:
> So, can topLevelObjCExposedDeclsInFileUnit contain global functions? If so, then you might see compound names here.

I don't see why not, I'll write a test to check.

Reply:
> So, can topLevelObjCExposedDeclsInFileUnit contain global functions? If so, then you might see compound names here.

Upon further investigation -- no. Only declarations with an @objc attribute appear in that list and as it turns out, you can't @objc annotate a Swift function.

```
test.swift:1:2: error: @objc can only be used with members of classes, @objc protocols, and concrete extensions of classes
@objc public func compoundNameTopLevelDecl(foo foo: Int, bar bar: Int) {
~^~~~~
```
So the declarations Swift name has to be simple. I also tried if it is for some reason possible to provide a Swift declaration with a compound @objc name, but that fortunately also isn't allowed:

```
test.swift:3:7: error: '@objc' class must have a simple name
@objc(my:compound:name:) public class Foo : NSObject {
      ^ ~~~~~~~~~~~~~~~
test.swift:9:7: error: '@objc' protocol must have a simple name
@objc(my:other:compound:name:) public protocol Bar {}
      ^ ~~~~~~~~~~~~~~~~~~~~~
test.swift:12:7: error: '@objc' extension must have a simple name
@objc(my:last:compound:name:) public extension Foo {
      ^ ~~~~~~~~~~~~~~~~~~~~
```